cmake: Fix passing `APPEND_*FLAGS` to `secp256k1` subtree

hebasto opened the pull request:
On the master branch @ 70e20ea024ce4f39abc4022e1ba19d5a6db2a207, the `APPEND_CPPFLAGS`, `APPEND_CFLAGS`  and `APPEND_LDFLAGS` are not correctly applied when building C code in the  `secp256k1` subtree, as intended.

This behaviour occurs due to two issues:
1. The command here: https://github.com/bitcoin/bitcoin/blob/70e20ea024ce4f39abc4022e1ba19d5a6db2a207/src/CMakeLists.txt#L77
does not affect the code in `add_subdirectory(secp256k1)` above it.

2.  `APPEND_LDFLAGS`  is not passed to the subtree's build system at all.

This PR fixes both issues.

Additionally, the helper variables `core_sanitizer_cxx_flags` and `core_sanitizer_linker_flags` have been removed.

dergoegge commented:
How can the broken behavior be observed? It appears to be working to me based on: the MSan job not spitting out false positives and coverage reports including coverage for secp.

fanquake commented:
> APPEND_LDFLAGS is not passed to the subtree's build system at all.

I thought #30791 fixed an issue related to passing down linker flags via `APPEND_LDFLAGS`. So has that regressed / broken since, or just never worked?

hebasto commented:
> How can the broken behavior be observed?

For example:
```
cmake -B build -DAPPEND_CFLAGS="-fprofile-update=prefer-atomic"
# Note the absence of the `SECP256K1_APPEND_CFLAGS ...` line in the secp256k1 summary
cmake --build build --target secp256k1 --verbose 
```

> It appears to be working to me based on: the MSan job not spitting out false positives and coverage reports including coverage for secp.

Yes, passing sanitizers flags works on the master branch.

**Another use case to consider:**
When building for test coverage with GCC the `-fprofile-update=prefer-atomic` flag has to be passed to a compiler on some build platforms. Using `-DAPPEND_CFLAGS="-fprofile-update=prefer-atomic"` does not work. 

> > APPEND_LDFLAGS is not passed to the subtree's build system at all.
> 
> I thought #30791 fixed an issue related to passing down linker flags via `APPEND_LDFLAGS`.

#30791 fixed passing the sanitizer flags.

>  So has that regressed / broken since, or just never worked?

The regression happened somewhere in the CMake staging branch. In this repository, it has never worked since #30454.

dergoegge commented:
> Yes, passing sanitizers flags works on the master branch.

Our MSan ci job [passes C/CXXFLAGS to depends](https://github.com/bitcoin/bitcoin/blob/efdb49afb9e24962958ad4445458ad8da253183b/ci/test/00_setup_env_native_msan.sh#L18) (including non-"-fsanitize=" flags e.g. `-fno-omit-frame-pointer -g -O1 -fno-optimize-sibling-calls`), why do those flags propagate to secp but the `APPEND_CFLAGS` flags do not? Would it make sense to use the same mechanism for both?

(I'm very confused by all the different ways of passing `*FLAGS`...)

theuni commented:
🤦🤦

Could you split this into 2 commits? The fixes you describe sound simple enough, but the removal of the helpers makes this hard to review.

hebasto committed:
cmake: Remove `core_sanitizer_{cxx,linker}_flags` helper variables

This change make the code more concise and minimizes the diff in the
subsequent commit.

hebasto committed:
cmake: Fix passing `APPEND_*FLAGS` to `secp256k1` subtree

hebasto commented:
> 🤦🤦
> 
> Could you split this into 2 commits? The fixes you describe sound simple enough, but the removal of the helpers makes this hard to review.

Sure thing! Done.

fanquake commented:
> I guess I'm a little confused about how this should work. It seems it's not very consistent.

> So.. how should this all work ideally? Do we really need to forward the flags at all?

I agree. It'd be good to know what the goals are here, if some amount of this code is redundant, and what the status of this change is.

hebasto commented:
> secp is the only C code we have. So arguably there's no need for `APPEND_CFLAGS` at all when `SECP256K1_APPEND_CFLAGS` can just be used directly, no?

Correct. It is solely about maintaining consistency with the other `APPEND_*` flags. I believe the following
```sh
cmake -B build -DAPPEND_CXXFLAGS="-fprofile-update=prefer-atomic" -DAPPEND_CFLAGS="-fprofile-update=prefer-atomic"
```
is better than
```sh
cmake -B build -DAPPEND_CXXFLAGS="-fprofile-update=prefer-atomic" -DSECP256K1_APPEND_CFLAGS="-fprofile-update=prefer-atomic"
```

> And as for `APPEND_LDFLAGS`, because secp is static, we don't actually invoke the linker for the library. Only for tests I guess?

Correct. Please refer to https://github.com/bitcoin-core/secp256k1/pull/1600.

> So.. how should this all work ideally? Do we really need to forward the flags at all? Asking because if no, that would explain how this has managed to be silently "broken".

Some flags, such as `-fprofile-update=prefer-atomic`, must be applied when building every object file. Additionally, flags like `-fsanitize=memory` must be applied during both compilation and linking.

By default, the main build system is configured without engaging any of these flags.

As [previously mentioned](https://github.com/bitcoin/bitcoin/pull/31379#issuecomment-2503562852), MSan-related flags are used without issue. This is because they are passed as `CFLAGS` to the depends build subsystem and subsequently to the main build system via a toolchain file.

Therefore, the issue has managed to remain silently _broken_ because there are no use cases in our tests.

Building for code coverage on some systems requires the `-fprofile-update=prefer-atomic` flag for both C++ and C code. This clearly exposes the bug.

theuni reviewed (approved):
utACK c4c5cf174883cb53256e869f0d1673e29576a97c.

It makes sense to forward `APPEND_LDFLAGS`  and `APPEND_CFLAGS` to their secp equivalents for convenience and consistency for tests.

sedited reviewed (approved):
ACK c4c5cf174883cb53256e869f0d1673e29576a97c

I think I prefer this over passing through `SECP256K1_APPEND_FLAGS`. This is a distinct project after all, and having to configure all its subtrees/subprojects with separate flags seems cumbersome. It would also be inconsistent, since the method introduced here for libsecp already works fine for e.g. leveldb.
